llama_stack/providers/inline/agents/meta_reference/agent_instance.py

```diff
@@ -246,6 +247,129 @@ async def create_and_execute_turn(self, request: AgentTurnCreateRequest) -> Asyn
 
             yield chunk
 
+    async def resume_turn(self, request: AgentTurnResumeRequest) -> AsyncGenerator:
+        with tracing.span("resume_turn") as span:
+            span.set_attribute("agent_id", self.agent_id)
+            span.set_attribute("session_id", request.session_id)
+            span.set_attribute("turn_id", request.turn_id)
+            span.set_attribute("request", request.model_dump_json())
+            assert request.stream is True, "Non-streaming not supported"
+
+            session_info = await self.storage.get_session_info(request.session_id)
+            if session_info is None:
+                raise ValueError(f"Session {request.session_id} not found")
+
+            turns = await self.storage.get_session_turns(request.session_id)
+
+            messages = []
+            if self.agent_config.instructions != "":
+                messages.append(SystemMessage(content=self.agent_config.instructions))
+
+            for i, turn in enumerate(turns):
+                messages.extend(self.turn_to_messages(turn))
+
+            messages.extend(request.tool_responses)
+
+            # get the steps from the turn id
+            steps = []
+            if len(turns) > 0:
+                steps = turns[-1].steps
+
+            # mark tool execution step as complete
+            in_progress_tool_call_step = await self.storage.get_in_progress_tool_call_step(
+                request.session_id, request.turn_id
+            )
+            tool_execution_step = ToolExecutionStep(
+                step_id=(in_progress_tool_call_step.step_id if in_progress_tool_call_step else str(uuid.uuid4())),
```

Comment:
shouldn't you error out if there is no step found?

Reply:
Agree its a bit confusing (had to play around with the react_agent app). Let me add a comment here. 

We do not error out here b/c in the case of ReActAgent (with a custom tool parser), server do not output a tool_execution step_start, and don't have the step. However, we should still allow the turn to be resumed with the ToolCallResponse in this case because even though server do not output a tool execution step.

Reply:
@yanxi0830 hm yeah this is confusing and broke my mental model.

Reply:
in that case, the server wouldn't have sent a `step_start` event either

Reply:
yeah, in this case (of custom tool parsers), the server wouldn't have sent a `step_start` event, but we would still like to log the ToolExecutionStep as a client tool call did indeed happen.